design/one-pager-default-resource-class.md

```diff
@@ -88,66 +119,308 @@ spec:
+   should check that the portable class's `classRef` references the correct
+   strongly typed class `kind`.
+
+In this model, the MySQL `RDSInstance` claim reconciler would be updated from
+its current state, which looks like this:
+
+```go
+// SetupWithManager adds a controller that reconciles MySQLInstance instance claims.
+func (c *MySQLInstanceClaimController) SetupWithManager(mgr ctrl.Manager) error {
+  r := resource.NewClaimReconciler(mgr,
+    resource.ClaimKind(databasev1alpha1.MySQLInstanceGroupVersionKind),
+    resource.ClassKind(v1alpha1.RDSInstanceClassGroupVersionKind),
+    resource.ManagedKind(v1alpha1.RDSInstanceGroupVersionKind),
+    resource.WithManagedConfigurators(
+      resource.ManagedConfiguratorFn(ConfigureMyRDSInstance),
+      resource.NewObjectMetaConfigurator(mgr.GetScheme()),
+    ))
+
+  name := strings.ToLower(fmt.Sprintf("%s.%s", databasev1alpha1.MySQLInstanceKind, controllerName))
+
+  return ctrl.NewControllerManagedBy(mgr).
+    Named(name).
+    Watches(&source.Kind{Type: &v1alpha1.RDSInstance{}}, &resource.EnqueueRequestForClaim{}).
+    For(&databasev1alpha1.MySQLInstance{}).
+    WithEventFilter(resource.NewPredicates(resource.HasClassReferenceKind(resource.ClassKind(v1alpha1.RDSInstanceClassGroupVersionKind)))).
+    Complete(r)
+}
+```
+To look as follows:
+
+```go
+// SetupWithManager adds a controller that reconciles MySQLInstance instance claims.
+func (c *MySQLInstanceClaimController) SetupWithManager(mgr ctrl.Manager) error {
+  r := resource.NewClaimReconciler(mgr,
+    resource.ClaimKind(databasev1alpha1.MySQLInstanceGroupVersionKind),
+    resource.ClassKinds{Portable: databasev1alpha1.MySQLInstanceGroupVersionKind, NonPortable: v1alpha1.RDSInstanceClassGroupVersionKind},
+    resource.ManagedKind(v1alpha1.RDSInstanceGroupVersionKind),
+    resource.WithManagedConfigurators(
+      resource.ManagedConfiguratorFn(ConfigureMyRDSInstance),
+      resource.NewObjectMetaConfigurator(mgr.GetScheme()),
+    ))
+
+  name := strings.ToLower(fmt.Sprintf("%s.%s", databasev1alpha1.MySQLInstanceKind, controllerName))
+
+  return ctrl.NewControllerManagedBy(mgr).
+    Named(name).
+    Watches(&source.Kind{Type: &v1alpha1.RDSInstance{}}, &resource.EnqueueRequestForClaim{}).
+    For(&databasev1alpha1.MySQLInstance{}).
+    WithEventFilter(resource.NewPredicates(resource.HasClassReferenceKind(mgr.GetClient(), resource.ClassKinds{Portable: databasev1alpha1.MySQLInstanceGroupVersionKind, NonPortable: v1alpha1.RDSInstanceClassGroupVersionKind}))).
+    Complete(r)
+}
 ```
 
-And a resource claim that defaulted to the resource class above could look like this in the general case:
+A claim referencing a portable class will now look as follows:
 
 ```yaml
-apiVersion: storage.crossplane.io/v1alpha1
-kind: NoSQLInstance
+apiVersion: database.crossplane.io/v1alpha1
+kind: MySQLInstance
 metadata:
-  name: dynamo-claim
-  namespace: demo
+  name: mysql-claim
 spec:
-  ...
+  classRef:
+    name: standard-mysql
```

Comment:
can this become `className:` instead of `classRef: name:`

Reply:
Hmm, this may be tricky. It would involve changing the embedded [`ResourceClaimSpec` in `crossplane-runtime`](https://github.com/crossplaneio/crossplane-runtime/blob/11c0da759c8181fee9926684654c1baa964ea421/apis/core/v1alpha1/resource.go#L49) which embeds `corev1.ObjectReference`. We are going to be [changing this](https://github.com/crossplaneio/crossplane/pull/743/files/7ac347caed82f479f8c971a103608716628dc5b2#diff-ac7b966ea5cff0d48966e06ad20a0106) to `corev1.LocalObjectReference`, which only has a `name` field. If we were to make this `inline`, the field would just be called `name`, not `className`. I am sure we could do some trickery to have a string field called `className` then create a `LocalObjectReference` from it, but the `Spec` would be less familiar to Kubernetes users and we would not be following common practice.

Reply:
I don't think it should block the design being merged, but wold be great to avoid naming the `group/version` given the 1:1 mapping between claim and class in this design.

Reply:
Oh yes, it will only accept `name` as it will now be `LocalObjectReference`

Reply:
Coming in late here, but either `className` or `classRef` using a `LocalObjectReference` (with only a `name` field) seem to be blessed by the Kubernetes API conventions. I think `LocalObjectReference` is a little more consistent with the rest of our API (we don't currently have any `fooName` fields), but we would benefit from a general API overhaul in this area; we have plenty of `LocalObjectReference` types that could be simple `fooName` fields, and I believe we may have a few `ObjectReference` fields that could be `LocalObjectReference`.